source/common/ssl/tls_certificate_config_impl.cc

```diff
@@ -64,19 +64,42 @@ TlsCertificateConfigImpl::TlsCertificateConfigImpl(
           fmt::format("Certificate configuration can't have both pkcs12 and private_key_provider"));
     }
   } else {
-    if (config.has_private_key_provider()) {
+    if (config.has_private_key_provider_list()) {
+      for (const auto& provider : config.private_key_provider_list().private_key_provider()) {
+        private_key_method_ =
+            factory_context.sslContextManager()
+                .privateKeyMethodManager()
+                .createPrivateKeyMethodProvider(provider, factory_context,
+                                                const_cast<std::string&>(private_key_));
+        if (private_key_method_ && private_key_method_->checkInitialized()) {
+          break;
+        } else {
+          private_key_method_ = nullptr;
```

Comment:
We should raise exception if the factory can't be found

Reply:
> We should raise exception if the factory can't be found

The overall exception if there is no any factory can be found is at this
```
if (config.has_private_key_provider_list()) {
        throw EnvoyException(fmt::format("Failed to load private key provider list."));
```

Reply:
it seems we won't reach that check when the fallback is true
`if ((private_key_.empty() || !fallback) && private_key_method_ == nullptr) {`

Reply:
done, thanks!